prog/test/hetzner_server.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+require "net/ssh"
+
+class Prog::Test::HetznerServer < Prog::Base
+  def self.assemble(server_id: Config.ci_hetzner_sacrificial_server_id)
+    if server_id.nil? || server_id.empty?
+      fail "server_id must be a nonempty string"
+    end
+    Strand.create_with_id(
+      prog: "Test::HetznerServer",
+      label: "start",
+      stack: [{server_id: server_id}]
+    )
+  end
+
+  label def start
+    hop_fetch_hostname
+  end
+
+  label def fetch_hostname
+    current_frame = strand.stack.first
+    current_frame["hostname"] = hetzner_api.get_main_ip4
+    strand.modified!(:stack)
+    strand.save_changes
+
+    hop_add_ssh_key
+  end
+
+  label def add_ssh_key
+    keypair = SshKey.generate
+
+    hetzner_api.add_key("ubicloud_ci_key_#{strand.ubid}", keypair.public_key)
+
+    current_frame = strand.stack.first
+    current_frame["hetzner_ssh_keypair"] = Base64.encode64(keypair.keypair)
+    strand.modified!(:stack)
+    strand.save_changes
+
+    hop_reset
+  end
+
+  label def reset
+    hetzner_api.reset(
+      frame["server_id"],
+      hetzner_ssh_key: hetzner_ssh_keypair.public_key
+    )
+
+    hop_wait_reset
+  end
+
+  label def wait_reset
+    begin
+      rootish_ssh("echo 1", hetzner_ssh_keypair.private_key)
+    rescue
+      nap 15
+    end
+
+    hop_setup_host
+  end
+
+  label def setup_host
+    st = Prog::Vm::HostNexus.assemble(
+      frame["hostname"],
+      provider: "hetzner",
+      hetzner_server_identifier: frame["server_id"]
+    )
+
+    current_frame = strand.stack.first
+    current_frame["vm_host_id"] = st.id
+    strand.modified!(:stack)
+    strand.save_changes
+
+    # invalidate frame so it's reloaded for subesequent statements in this method
+    @frame = nil
```

Comment:
I think you should consider overriding the `frame` method altogether. More on that in the main PR comments...

Reply:
This is not really necessary here, instead I changed the code to fetch the sshable by doing `Sshable[st.id]` directly instead of depending on the `sshable` method.